src/System.CommandLine/Parsing/ParseResultVisitor.cs

```diff
@@ -153,88 +153,78 @@ protected override void Stop(SyntaxNode node)
+                        ValidateArgumentResult(argumentResult);
 
-                        case OptionResult optionResult:
-
-                            var argument = optionResult.Option.Argument;
+                        break;
 
-                            var arityFailure = ArgumentArity.Validate(
-                                optionResult,
-                                argument,
-                                argument.Arity.MinimumNumberOfValues,
-                                argument.Arity.MaximumNumberOfValues);
+                    case OptionResult optionResult:
 
-                            if (arityFailure != null)
-                            {
-                                _errors.Add(
-                                    new ParseError(arityFailure.ErrorMessage, optionResult));
-                            }
+                        ValidateOptionResult(optionResult);
 
-                            var results = optionResult.Children
-                                                      .OfType<ArgumentResult>()
-                                                      .ToArray();
+                        break;
+                }
+            }
+        }
 
-                            foreach (var a in results)
-                            {
-                                ValidateArgument(a);
-                            }
+        private void ValidateCommandResult(CommandResult commandResult)
+        {
+            if (commandResult.Command is Command command)
+            {
+                foreach (var validator in command.Validators)
+                {
+                    var errorMessage = validator(commandResult);
 
-                            break;
+                    if (!string.IsNullOrWhiteSpace(errorMessage))
+                    {
+                        _errors.Add(
+                            new ParseError(errorMessage, commandResult));
                     }
                 }
             }
-        }
 
-        private void ValidateCommand(CommandResult commandResult)
-        {
-            foreach (var a in commandResult
-                              .Command
-                              .Arguments)
+            foreach (var option in commandResult
+                                   .Command
+                                   .Options)
             {
-                if (a is Argument argument)
+                if (option is Option o &&
+                    o.Required && 
+                    _rootCommandResult.FindResultFor(o) == null)
+                {
+                    _errors.Add(
+                        new ParseError($"Option '{o.RawAliases.First()}' is required.",
+                                       commandResult));
+                }
+            }
+
+            foreach (var symbol in commandResult
+                                   .Command
+                                   .Arguments)
+            {
+                if (symbol is Argument argument)
```

Comment:
This `if` check does not appear to add much. It prevents a `null` argument from getting through. But it is not possible to get a `null` back from `Command.Arguments` due to the implementation using `.OfType`. I would recommend just changing the foreach to be:
```C#
            foreach (IArgument argument in commandResult
                                   .Command
                                   .Arguments)
            {
```